libcudacxx/include/cuda/try_cancel

```diff
@@ -0,0 +1,324 @@
+ * restrictions in this license pursuant to DFARS 227.7202-3(a) or as set forth in subparagraphs (b)(1) and (2) of the
+ * Commercial Computer Software - Restricted Rights clause at FAR 52.227-19, as applicable. Contractor/manufacturer is
+ * NVIDIA, 2788 San Tomas Expressway, Santa Clara, CA 95051.
+ *
+ * 15. ENTIRE AGREEMENT. This license is the final, complete and exclusive agreement between the parties relating to the
+ * subject matter of this license and supersedes all prior or contemporaneous understandings and agreements relating to
+ * this subject matter, whether oral or written. If any court of competent jurisdiction determines that any provision of
+ * this license is illegal, invalid or unenforceable, the remaining provisions will remain in full force and effect.
+ * This license may only be modified in a writing signed by an authorized representative of each party.
+ *
+ * (v. August 20, 2021)
+ */
+
+#ifndef _CUDA_TRY_CANCEL
+#define _CUDA_TRY_CANCEL
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+namespace experimental {
+
+namespace __detail {
+
+struct __empty_t {
+  __device__ void operator()(dim3);
+};
+
+template <int __I>
+__device__ int __cluster_get_dim(__int128 __result) {
+    int __r;
+    if constexpr(__I == 0) {
+        asm volatile("clusterlaunchcontrol.query_cancel.get_first_ctaid::x.b32.b128 %0, %1;" : "=r"(__r) : "q"(__result) : "memory");
+    } else if constexpr(__I == 1) {
+        asm volatile("clusterlaunchcontrol.query_cancel.get_first_ctaid::y.b32.b128 %0, %1;" : "=r"(__r) : "q"(__result) : "memory");
+    } else if constexpr(__I == 2) {
+        asm volatile("clusterlaunchcontrol.query_cancel.get_first_ctaid::z.b32.b128 %0, %1;" : "=r"(__r) : "q"(__result) : "memory");
+    } else {
+      __r = -1;
+      #ifndef NDEBUG
+      __trap();
+      #endif
+    }
+    return __r;
+}
+
+} // namespace __detail
+
+/// This API for implementing work-stealing, repeatedly attempts to cancel the launch of a thread block
+/// from the current grid. On success, it invokes the unary function `__uf` before trying again.
+/// On failure, it returns.
+///
+/// This API does not provide any memory synchronization.
+/// This API does not guarantee that any thread will invoke `__uf` with the next block index until all
+/// invocatons of `__uf` for the prior block index have returned.
+///
+/// Preconditions:
+/// - All thread block threads shall call this API exactly once.
+/// - Exactly one thread block thread shall call this API with `__is_leader` equals `true`.  
+template <int __ThreadBlockDim = 3, typename __UnaryFunction = __detail::__empty_t>
+    requires std::is_invocable_r_v<void, __UnaryFunction, dim3>
```

Comment:
We are still supporting C++17 so 
```suggestion
_CCCL_TEMPLATE(int __ThreadBlockDim = 3, typename __UnaryFunction = __detail::__empty_t)
_CCCL_REQUIRES(_CUDA_VSTD::is_invocable_r_v<void, __UnaryFunction, dim3>)
```

Reply:
Have added a C++20 guard for now, but leaving the thread open to provide a full C++17 backport when we are happy with the current API design.